### ISSUE TITLE
Potential improvements for evaluation

### ISSUE THREAD

**Issue by BenjaminBossan** [pratyushasharma/laser#15]
Thanks for providing the code for this promising research. I'm looking forward to see how far this idea can be pushed. It would be especially cool if a heuristic could be found that applies this technique to multiple layers chosen in a way that works across different models.

When I investigated the results a bit closer and ran some of the benchmarks locally, I came across some potential issues. Specifically, I took a look at the BigBench-Epistemic Reasoning benchmark, but I suspect that others could also be affected. First of all, I noticed that the accuracy of the models without intervention was below 50% (Tab. 1). For a binary classification task, this is strange. When debugging the results, I found that for Roberta and GPT-J (haven't tested Llama), the models always predicted the same label, and since that label was used in 37% of samples in the dataset, that's also their accuracy. As Llama has 63% accuracy with intervention, I suspect that it simply always predicts other label.

Digging a bit deeper, I found the logits for the label tokens to be extremely small. This typically happens when the model is somehow "derailed" and wants to predict neither of the tokens. Sometimes, this simply comes down to tokenization: Often, the models try to predict " True" and " False" (leading whitespace) because this is how they tokenize the text. Other times, they want to go in a completely different direction. I would recommend to log the absolute probabilities of the label tokens and double-check when they are too low. Often, this can be fixed by slight adjustments to prompts or labels.

Also, there is a typo in [this prompt](https://github.com/pratyushasharma/laser/blob/7e8a2550e7b0dbc1fe466b82aba7b9a791bdb5c0/src/dataset_utils/bigbench.py#L57): "entails" => "entail".

I hope this is helpful.

**Comment by dkmisra** [comment]
Hi @BenjaminBossan . Thanks for these comments. 

Regarding your first comment, we have a stacking result in our paper where we apply these interventions layer by layer in a greedy fashion.
See the paragraph on `Composing reductions across layers` in the paper [here](https://arxiv.org/pdf/2312.13558.pdf).

Your second comment sounds very important and interesting. I am logging the predicted labels and log-prob but I haven't inspected them manually. We have a few binary classification tasks: Fever, Bios Gender, TruthfulQA and BigBench Epistemic Reasoning. I'll log their uniform prediction accuracy and paste it here. I see that Bios Gender does reach 97% and it is unlikely that 97% of the labels at one gender, so maybe it is limited to Epistemic Reasoning - perhaps the models cannot solve the task and instead the optimal intervention just shifts the bias. Let me check and get back today.

Thanks for the fix in the prompt typo. I'll make a note of it. 

The issue of whitespace before tokens is also important. I made changes so that GPTJ predicts token with a whitespace and Llama2 doesnt but I will do another sanity check.
